app/models/gather_repository_maintenance_stats.rb

```diff
@@ -0,0 +1,108 @@
+class GatherRepositoryMaintenanceStats
+    def self.gather_stats(repository)
+        return unless repository.host_type == "GitHub" # only support Github repos for now
+        client = AuthToken.v4_client
+        v3_client = AuthToken.client
+        stats_to_run = get_stats_to_run(repository)
+
+        clients = {
+            v3: v3_client,
+            v4: client
+        }
+
+        metrics = []
+
+        stats_to_run.each do |stat_run|
+            query_to_run = stat_run[:query].new(clients[stat_run[:query].client_type])
+            begin
+                result = query_to_run.query(params: stat_run[:variables])
+                unless check_for_error_response(result, stat_run[:query].client_type)
+                    stat_run[:stat_class].each do |stat_class|
+                        metrics << stat_class.new(result).get_stats
+                    end
+                end
+            rescue Octokit::Error
+                next
+            end
+        end
+
+        add_metrics_to_repo(repository, metrics)
+
+        metrics
+    end
+
+    private
+
+    def self.check_for_error_response(response, client_version)
+        if client_version == :v4
+            # errors can be stored in the response from Github or can be stored in the response object from HTTP errors
+            response.errors.each do |message|
+                Rails.logger.warn(message)
+            end
+            response.data.errors.each do |message|
+                Rails.logger.warn(message)
+            end unless response.data.errors.nil?
+            # if we have either type of error or there is no data return true
+            return response.data.nil? || response.errors.any? || response.data.errors.any?
+        end
+        # return false for v3 for right now since those raise errors
+        false
+    end
+
+    def self.add_metrics_to_repo(repository, results)
+        # create one hash with all results
+        results.reduce(Hash.new, :merge).each do |category, value|
+            unless value.nil?
+                stat = repository.repository_maintenance_stats.find_or_create_by(category: category.to_s)
+                stat.update_attributes!(value: value.to_s)
+                # force rails to set updated_at
+                stat.touch
```

Comment:
Why not just do
```ruby
stat.update!(value: value.to_s)
```

which then also does the updating of `updated_at` as it saves?

Reply:
I wanted to make sure the `updated_at` value changes even if the stat did not change. Right now I'm planning on using the stats `updated_at` field to help figure out prioritization for refreshing repositories.

Reply:
In that case, it's probably still worth doing something like
```
stat.update!(value: value.to_s)
stat.touch unless stat.changed?  # we always want to update updated_at for later querying
```
to only do one db call for the changed case